tree_arena/src/tree_arena_unsafe.rs

```diff
@@ -0,0 +1,592 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#![allow(unsafe_code, reason = "Purpose is unsafe abstraction")]
+use super::NodeId;
+
+use std::cell::UnsafeCell;
+use std::collections::HashMap;
+
+#[derive(Debug)]
+struct TreeNode<T> {
+    item: T,
+    children: Vec<NodeId>,
+}
+
+/// Mapping of data for the Tree Arena
+#[derive(Debug)]
+struct DataMap<T> {
+    /// The items in the tree
+    items: HashMap<NodeId, Box<UnsafeCell<TreeNode<T>>>>,
```

Comment:
Would we be able to have something like an `UnsafeMutex` type, ideally which is implemented safely by default but with an unsafe passthrough version, to check some that invariants are being met.
That is, have this same implementation, but in a checked manner?

Reply:
Yes but would require quite a few changes as the TreeNode contains the contents themselves and the list of children, but the API returns these in separate structs - the` item` and the `ArenaChildren` (mut or ref) so I think both would  need to be wrapped in some form of mutex for run time checking separately ?